Symantec Endpoint Security UDI connector

thangaraj-ramesh opened the pull request:
Symantec Endpoint Security UDI connector

thangaraj-ramesh committed:
Symantec Endpoint Security UDI connector

Symantec Endpoint Security UDI connector

thangaraj-ramesh committed:
Merge branch 'opencybersecurityalliance:develop' into symantec_endpoint_security

thangaraj-ramesh committed:
Updated transmit code and minor code obfuscation fixes

Minor updated to transmit metadata and small fixes for query with path

DerekRushton commented on stix_shifter_modules/symantec_endpoint_security/stix_transmission/connector.py:
```diff
@@ -0,0 +1,278 @@
+        response_dict = {}
+        return_obj = {}
+        response = response_wrapper.read().decode('utf-8')
+        if response.startswith('{') and response.endswith('}'):
```
No need to have this. You can perform the load and if it fails, just catch the exception and continue. Checking just the start/end may not mean it's a valid json object.

thangaraj-ramesh replied:
Updated code as per suggestion.

DerekRushton commented on stix_shifter_modules/symantec_endpoint_security/stix_transmission/connector.py:
```diff
@@ -0,0 +1,278 @@
+        if response_wrapper.code != 200:
+            response_dict['message'] = response_dict.get('message', '')
+            if response_dict.get('fault'):
+                response_dict['message'] = response_dict.get('fault', '').get('faultstring', '')
```
The second .get() on this line may cause some issues if it fails to resolve. Can you split this into two lines and do a check on the second line to ensure the first object resolved?

fault = response_dict.get('fault', '')
if (fault != '')
fault.get('faultstring', '')

thangaraj-ramesh replied:
'If 'fault' isn't present in response_dict, then if block doesn't get executed since get() will return None.
If 'fault' is present in response_dict and 'faultstring' isn't present then an empty string is returned by second get(). Hence no change in code was required.

DerekRushton commented on stix_shifter_modules/symantec_endpoint_security/stix_transmission/connector.py:
```diff
@@ -0,0 +1,278 @@
+            query['next'] = offset
+
+            # Update query start_date and next values from metadata to get more than 10k results.
+            if metadata:
```
This shouldn't be needed, you already check in the method.

thangaraj-ramesh replied:
Updated code as per suggestion.

DerekRushton commented on stix_shifter_modules/symantec_endpoint_security/stix_transmission/connector.py:
```diff
@@ -0,0 +1,278 @@
+        """
+        if metadata:
+            if isinstance(metadata, dict) and metadata.get('start_date_event_count') and metadata.get('start_date'):
+                query['next'] = offset % SYMANTEC_MAX_QUERY_RESULTS + int(metadata.get('start_date_event_count', 0))
```
Is it possible for the offset to be higher than the "SYMANTEC_MAX_QUERY_RESULTS"?

thangaraj-ramesh replied:
Yes, the value for "SYMANTEC_MAX_QUERY_RESULTS"  is set to 10000 as it is a limitation of API. Symantec UDI connector is supporting results limit more than the API limitation by creating new query. Hence offset higher than the limit "SYMANTEC_MAX_QUERY_RESULTS" can be passed based on result limit.

DerekRushton commented on stix_shifter_modules/symantec_endpoint_security/stix_translation/json/from_stix_map.json:
```diff
@@ -0,0 +1,256 @@
+  },
+  "mac-addr": {
+    "fields": {
+      "value": ["device_mac", "device_networks.mac", "device_networks.gateway_mac"]
```
Should [device_networks.gateway_mac] be included under the resolves_to_ref?

thangaraj-ramesh replied:
Updated code as per suggestion.

DerekRushton commented on stix_shifter_modules/symantec_endpoint_security/stix_translation/json/from_stix_map.json:
```diff
@@ -0,0 +1,256 @@
+      "command_line": ["actor.cmd_line","process.cmd_line", "parent.cmd_line", "startup_app.cmd_line"],
+      "name": ["actor.app_name", "process.app_name", "parent.app_name"],
+      "created": ["actor.start_time", "process.start_time", "parent.start_time"],
+      "x_thread_id": ["actor.tid", "process.tid"],
```
Given these are custom fields, I'd rename them to just x_tid and x_uid.

thangaraj-ramesh replied:
As recommended by STIX naming standards, "Cyber Observable Object custom property should start with 'x_' followed by a source unique identifier (like a domain name with dots replaced by hyphens), a hyphen and then the name.". This is the reason custom attributes are named with minimum two hyphens .

DerekRushton replied:
Could you call it x_process_tid and x_process_uid? I just want to keep them labeled as they appear in the API response.

thangaraj-ramesh replied:
Updated as per suggestion.

DerekRushton commented on stix_shifter_modules/symantec_endpoint_security/stix_translation/json/from_stix_map.json:
```diff
@@ -0,0 +1,256 @@
+      "hashes.'SHA-1'": ["file.sha1", "actor.file.sha1", "parent.file.sha1", "process.file.sha1", "startup_app.file.sha1"],
+      "created": ["actor.file.created", "parent.file.created", "process.file.created", "startup_app.file.created"],
+      "modified": ["actor.file.modified", "process.file.modified", "startup_app.file.modified"],
+      "x_file_type": ["file.type_id", "actor.file.type_id", "process.file.type_id"],
```
x_file_typeid or x_type_id or x_typeid

thangaraj-ramesh replied:
The attribute value in stix output is transformed to string value and for query also string value is supported. Hence "id" isn't used in the name. For example ; x_file_type= "Directory" gets transformed to  file.type_id= 2.

DerekRushton replied:
Makes sense.

DerekRushton commented on stix_shifter_modules/symantec_endpoint_security/stix_translation/json/from_stix_map.json:
```diff
@@ -0,0 +1,256 @@
+      "x_rep_score": ["file.rep_score"],
+      "x_file_version": ["file.version"],
+      "x_open_mode": ["open_mode"],
+      "x_signature_company_name": ["actor.file.signature_company_name", "file.signature_company_name", "module.signature_company_name", "parent.file.signature_company_name", "process.file.signature_company_name"],
```
These signature fields seem like they are related to a certificate. If they are would it make sense to log them under ​X.509 Certificate Object

thangaraj-ramesh replied:
Added mapping for these fields to x509-certificate object.

DerekRushton commented on stix_shifter_modules/symantec_endpoint_security/stix_translation/json/from_stix_map.json:
```diff
@@ -0,0 +1,256 @@
+      "x_signature_serial_number": ["actor.file.signature_serial_number"],
+      "x_signature_value": ["actor.file.signature_value"],
+      "x_signature_value_ids": ["actor.file.signature_value_ids", "process.file.signature_value_ids", "startup_app.file.signature_value_ids"],
+      "x_content_type": ["file.content_type.type_id"]
```
Probably should have id in the name.

thangaraj-ramesh replied:
'type_id' is the general type of a file. Thus it is mapped to custom attribute 'file.x_content_type'

DerekRushton commented on stix_shifter_modules/symantec_endpoint_security/stix_translation/json/from_stix_map.json:
```diff
@@ -0,0 +1,256 @@
+      "name": ["device_os_name"],
+      "version": ["device_os_ver"],
+      "x_os_type": ["device_os_type_id"],
+      "x_cmd_line": ["startup_app.cmd_line"],
```
Does these extensions really fit under software?
x_cmd_line
x_start_type_id
x_config_path

thangaraj-ramesh replied:
Here Startup_app is mapped as software object. These attribues are related to startup_app. Removing it now from from_stix mapping .

DerekRushton commented on stix_shifter_modules/symantec_endpoint_security/stix_translation/json/from_stix_map.json:
```diff
@@ -0,0 +1,256 @@
+  "url": {
+    "fields": {
+      "value": ["url.text", "file.url.text"],
+      "x_url_host": ["url.host", "file.url.host", "connection.url.host"],
```
These three extensions are a bit odd. Are they here because they have a value that only make sense under an extension? These look like they could fit under another object.

thangaraj-ramesh replied:
These attribues are related extracted parts of URL text itself. Removing it now from from_stix mapping .

DerekRushton commented on stix_shifter_modules/symantec_endpoint_security/README.md:
```diff
@@ -0,0 +1,700 @@
+- The access token expires 60 minutes after being created.
+
+### Limitations
+- The maximum number of API calls for events is limited to 500 per hour.
```
Do we have an error message for this?

thangaraj-ramesh replied:
Yes, It is already handled by framework.

DerekRushton replied:
I'd like to see a specific check made in the code for this one. 500 calls an hour is relatively low and this could be a regular issue. I'd rather be certain it's caught.

thangaraj-ramesh replied:
It is handled by framework as follows: 
Error: 429 (Too many requests) is returned by data source server. This error is handled by framwork in function "async def call_api()" file "stix_shifter_utils/stix_transmission/utils/RestApiClientAsync.py"

Connector sends the following Error:
RuntimeError: Fetching results failed; see log for details{'success': False, 'connector': 'symantec_endpoint_security', 'error': 'symantec_endpoint_security connector error => Max retries exceeded. too_many_requests with max retry (1)', 'code': 'too_many_requests'}

DerekRushton replied:
The problem with that error message is that a user won't be able to easily identify why they are erroring out. Since we know that the API will give that message at 500 calls, we should be explicit in the output that "Too many request were made in the last hour. This API is limited to 500 request per an hour."

thangaraj-ramesh replied:
Added custom message for 429 error code as suggested.

DerekRushton commented on stix_shifter_modules/symantec_endpoint_security/README.md:
```diff
@@ -0,0 +1,700 @@
+| List of dictionary fields | A list containing one or more dictionaries | "device_networks": [{"ipv4": "1.1.1.1", "ipv6": "xx22::2x2x:22xx:22xx:x2x2", "mac": "11:1X:XX:X1:11:X1" }, { "ipv4": "2.2.2.2", "ipv6": "xx11::1x1x:11xx:11xx:x1x1", "mac": "22:X2:22:X2:X2:22"}] |
+
+### Observations
+- The access token expires 60 minutes after being created.
```
Do we have an error message for this.

thangaraj-ramesh replied:
This error won't occur in connector as new token is getting created for every API request. It is just as an observation and it is now removed from ReadMe.

thangaraj-ramesh committed:
Updated with review comments

Removed some extension from mapping and handled exception for invalid json response.

thangaraj-ramesh committed:
Merge branch 'develop' into symantec_endpoint_security

thangaraj-ramesh committed:
Updated code as per review comments

Updated code as per review comments:
- x509-certificate stix object added for signature related fields.
- Added custom messages for error handling.

DerekRushton committed:
Merge branch 'develop' into symantec_endpoint_security
